### DISCUSSION TITLE
Synchronising styles of Express Checkout Buttons

### DISCUSSION THREAD

**Discussion by alexflorisca** [woocommerce/woocommerce#48055] [WooCommerce, Blocks, and Site Editing]
I’ve recently been doing some research and a proof of concept for synchronising the button styles for all checkout buttons.

## TL;DR
Whist we control the layout where the express checkout buttons are rendered, the styling of each button is controlled by each payment extension. This can cause visual inconsistency in the checkout block, where buttons are rendered with different styles (height, rounded corners, label, etc). This discussion and associated POCs ([Woo](https://github.com/woocommerce/woocommerce/pull/47899), [WooPayments](https://github.com/Automattic/woocommerce-payments/pull/8888), [Stripe](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3164)), explores the current landscape of express payment buttons and what we can do to ensure cohesiveness across the express checkout area

This is a POC in Woo for passing down style information to express payment gateways:
[[POC: Pass style properties to Express Checkout buttons](https://github.com/woocommerce/woocommerce/pull/47899#top)](https://github.com/woocommerce/woocommerce/pull/47899#top)

And these two are the corresponding changes in Stripe and WooPayments to receive and respect these styles.
WooPayments - [[POC: Receive button properties from the Checkout Block](https://github.com/Automattic/woocommerce-payments/pull/8888#top)](https://github.com/Automattic/woocommerce-payments/pull/8888#top)
Stripe - [[POC: Receive button properties from the Checkout Block](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3164#top)](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3164#top)

I wanted to start a discussion to gather feedback on this approach and especially involve developers who build/maintain any payment extensions that interact with express payment methods.

## The ideal solution

The ideal solution would be where the payment extensions would simply need to provide a few props like the logo, button text, etc and we would render a React component that takes care of every button’s appearance.

The reality isn’t so simple…

## The current landscape

I looked at Stripe, WooPayments, PayPal and Amazon as they seem the most popular. I also found that quite a few payment plugins use the Stripe integration (WooPayments does this) for express payments which essentially uses the same API and renders the same button.

Below is a summary

- Each payment gateway is responsible for styling the buttons.
- Each payment gateway provides merchants with styling options such as button label, size, rounded corners and color. These differ between payment gateways.
- Most, but not all, payment gateways render the buttons inside iframes
- The top payment gateways (Stripe, Paypal, Amazon) have SDKs that are used across a wide range of products, their WooCommerce extensions being one of them
- Express payment buttons need to work on Cart & Checkout Blocks, Shortcode, Product and Settings pages.

## Restrictions

So we have a few restrictions to work with.

> Buttons are rendered inside iframes

This is partly for security and partly for controlling the button appearance. @pierorocca mentioned that brands like Apple and Google have certain requirements for sticking to brand guidelines wherever the buttons are used. Rendering the buttons in iframes means they have more control over this. It also means we can't really offer a generic component to the payment gateways to render.

While we did consider this approach for express payment methods that do not use iframes (such as the WooPay button), they are far and few between, and to me it makes sense to optimise for the majority rather than minority. I would also expect that accommodating the various changing needs of different payment extensions with one component could get tricky and time consuming with the API changing quite frequently.

> The styling options differ between payment gateways.

In my POC, I explored the idea of centralising styling settings within the Express Checkout Block. Unfortunately, every payment gateway accepts different settings, especially for the button label, so this won't work for every setting.

> Express payment buttons need to work on Cart & Checkout Blocks, Shortcode and Product pages.

This is also a problem because the buttons rendered on shortcode versions of Cart & Checkout won't have access to these centralised style attribues.

Below is a summary of the top payment gateways and their support for styling properties

|                 | Woo Pay                                        | Link               | Apple Pay                                                                                                                                                                                       | Google Pay                                                                                                     | PayPal                                       | Amazon                        |
|-----------------|------------------------------------------------|--------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------|----------------------------------------------|-------------------------------|
| Button Height   | ✔                                              | ✔                  | ✔                                                                                                                                                                                               | ✔                                                                                                              | ✔                                            | ?                             |
| Button Color    | Purple<br>Light Purple<br>Light Purple Outline | Green              | Black<br>White<br>White outline                                                                                                                                                                 | Black<br>White                                                                                                 | Blue<br>Silver<br>White<br>Black<br>Gold     | Gold<br>LightGrey<br>DarkGrey |
| Rounded Corners | ✔                                              | n/a<br>only in ECE | n/a<br>only in ECE                                                                                                                                                                              | n/a<br>only in ECE                                                                                             | Square<br>Pill                               | n/a                           |
| Button Label    | Logo only<br>Buy<br>Donate<br>Book             | Pay faster         | Logo only<br>Add Money with<br>Book with<br>Buy with<br>Check Out with<br>Contribute with<br>Donate with<br>Order with<br>Reload with<br>Rent with<br>Subscribe with<br>Tip with<br>Top up with | Logo only<br>Book with<br>Buy with<br>Checkout with<br>Donate with<br>Order with<br>Pay with<br>Subscribe with | Logo only<br>Checkout<br>Buy Now<br>Pay with | n/a                           |

[1] We already control the layout and the width of the buttons. They all render as 100% width of their container.
[2] Amazon Pay doesn't currently support `height` as a property, but it will respect the height of the container.
[3] ECE - [Express Checkout Elements](https://docs.stripe.com/elements/express-checkout-element) - Stripe's new API for rendering express buttons

## Labels
The button labels are work a special mention here as I've had several rounds of thoughts about these. Initially I thought it would be great to sync the button labels across buttons, however this isn't possible currently as they vary so wildly between payment gateways. 

I then thought maybe it's possible to work with the payment providers to sync on these but I quickly dismissed that as it would be a hard sell to ask them to change their APIs for us. Especially because of this:

> The top payment gateways (Stripe, Paypal, Amazon) have SDKs that are used across a wide range of products, their WooCommerce extensions being one of them

@ralucaStan made a good suggestion about removing the label altogether and just displaying the default button logo. @pierorocca pointed out a few scenarios where this won't work (single page donation pages). My thinking now is that it's best to encourage payment gateways to default to show no label (no text associated with the button like "Buy with"), just the logo. Merchants can then change their settings individually for each express payment method if they wish to do so. This is how it currently works, so there's not much change here, and it would provide a clean default experience with the option of customising. Something we're always aiming for with Cart & Checkout blocks.

## My proposal

Given all the above, at a high level this is what I'd propose

- Whatever styles we can centralise (e.g. button height), we do so by adding attributes in the express checkout block.
- We keep existing gateway settings, but if we are in a blocks context, we override them with centralised values from the Express Checkout block.
- We pull in style attributes from the current theme where applicable (e.g. border radius) to better sync with overall theme
- We add a `buttonAttributes` property to the current properties that are inherited by the express payment buttons.
- Each payment method can then use these when rendering their buttons. I have created PRs in [WooPayments](https://github.com/Automattic/woocommerce-payments/pull/8888) and [Stripe](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3164) to demonstrate how this works
- Encourage express payment buttons are rendered by default with no additional text, but can be edited if the merchant wishes to do so

Going in a bit more detail, this means
- Expose the button `height` as an option in the Express Checkout block
- Use the existing `hasDarkControls` and pass that down to the express payment button
  - The payment extension can then decide what do with that. Some might choose between light/dark more buttons, some might ignore it if providing only 1 color option. It gives them a bit more information about the current theme and what would look best.
- Get the `borderRadius` from the current theme settings for the round corner setting.
- Default to no label for express payment buttons

### Pros

- Improves cohesiveness of the express checkout area
- Supports all express payment methods
- Slightly more control to the merchant
- Able to respect theme settings
- Backwards compatible - payment extension specific settings will still apply to all non block contexts (settings pages, shortcode Cart &amp Checkout)

### Cons

- Payment gateway developers need to implement changes in their extensions
  - Although I would say the changes needed are fairly simple and easy to do
- Not guaranteed to be followed by everyone

## Summary
- We already control the layout and the width of the buttons
- `height` and `borderRadius` are the 2 properties that would have the biggest impact in terms of 
visual consistency for the checkout block. If we can get as many express payment buttons to use the same values for those, it would be a big win
- The `hasDarkInputs` property can provide additional information about the best button color to render that won't clash with the current background.

## Next steps

I would love to get feedback on this proposal, especially from any payment extension developers that integrate with Woo. If this is something we want to go ahead with, I will then
- communicate our planned approach via developer.woocommerce.com and WC Slack 
- create PRs in Amazon Pay and Paypal to make sure this works there too
- polish off the POC in Woo and merge
- Merge all the POC branches in the various payment gateway repositories and make sure the changes are released in a timely manner
- Create documentation that other payment gateway developers can use to apply these changes

## Questions
Tagging a few people below based on responsibility, contributions and who _I think_ would have a valuable input here. Please tag anyone else you think is relevant.

I'd like to know:
1. Does this approach work with your payment extension?
2. Could you add some logic to change the button color given the dark mode setting?
3. Do you see any problems or conflicts with this approach?
4. Is there a better solution?

@woocommerce/rubik @pierorocca @pmcpinto @nerrad @woocommerce/quark @woocommerce/fractal @woocommerce/axiom @woocommerce/heisenberg @dpanta94 @AlexP11223

**Comment by pierorocca** [comment]
Thanks for the exploration Alex. Focusing on things that are most common like height, styling/theme, and placement seems reasonable. Some cases like Amazon Pay, Link and PayPal, even with no text label specified the button is going to contain text or in GooglePay case brand and last 4 digits. No text as default is a good option imo.


Same with gateways like Afterpay, who additionally don't adhere to design guidelines. We've reached out to them to encourage compliance, which is our best bet at this time. We'll keep showing 3Ps the lack of interoperability and keep influencing iterations to their designs while on WooPayments we'll move to ECE which adds corner radius and Stripe plugin will move to ECE in 2025.

**Comment by frosso** [comment]
I like the idea of centralizing the common settings 👍 

> We keep existing gateway settings, but if we are in a blocks context, we override them with centralised values from the Express Checkout block.
> [...]
> Backwards compatible - payment extension specific settings will still apply to all non block contexts (settings pages, shortcode Cart &amp Checkout)

I'd like some clarification on the consequences of this decision.
If a merchant changed some settings from the extension's specific settings page, how would such change influence the button rendered in a block context?

I could see some merchants using a block-based cart or checkout, but still using a non-block theme (or vice versa). Would a merchant have to know they have to manually sync the settings for both experiences?

**Reply by alexflorisca** [reply]
> If a merchant changed some settings from the extension's specific settings page, how would such change influence the button rendered in a block context?

This is a good point. And I like "common settings", let's use that to refer to settings that apply to all payment methods, and that would be displayed as attributes on the express checkout block in the editor. Where there is a common setting (e.g. button size), it will override the extension specific setting within a block context. If the merchant changes an extension specific setting that doesn't have a common setting equivalent, then that would apply to the button in a block context as well.

> I could see some merchants using a block-based cart or checkout, but still using a non-block theme (or vice versa). Would a merchant have to know they have to manually sync the settings for both experiences?

Yeah this isn't ideal. We could display a notice for the common settings to say "hey these only apply in a block context. For all other settings, go to your payment extension's settings page". It's a bit of a tradeoff here with some added complexity for the merchant for a slicker experience for the shopper. Personally I think it's worth it, and I don't have any data to back this up, but I envisage for the majority of stores, it won't be necessary to customise the express checkout buttons individually if we are providing a great default experience (everything aligned in terms of layout and style).

I also had a thought that we could write some code to sync the settings from the block editor to let's say the top 5 most popular payment gateways, but we can't do it for every single one. And that smells pretty bad to me so I don't think it's a good idea to go down the route of "sync with some, but not others"

I'd be interested to hear if anyone has any other ideas here?

**Comment by pmcpinto** [comment]
Alex, thanks for doing this spike and proposing a path forward. Focusing on height and border-radius, plus defaulting to no label, is a good solution considering all the limitations.

Providing enough context in the editor and admin settings will be critical to mitigate some ambiguity created by this approach. Cc @elizaan36  @nikkivias in case you have some thoughts.

@InpsydeNiklas it would be great to have Syde's feedback on this proposal, as PayPal leverages the express checkout section. Thanks.

**Reply by InpsydeNiklas** [reply]
Hello @pmcpinto @alexflorisca, as discussed, we have created a draft PR for PayPal Payments here: https://github.com/woocommerce/woocommerce-paypal-payments/pull/2354
Passing the configuration from the Checkout block is not a major difficulty. However, one challenge is the defined [standard for the button height](https://developer.woocommerce.com/docs/user-experience-guidelines-payment-button-size-and-anatomy/). Because the `Large` style is defined with `56px`, but PayPal only supports up to `55px` for the buttons (Apple Pay & Google Pay buttons are more flexible).

Beyond this, what I am missing from the proposal is a general guideline on how the payment gateway configuration should communicate with users that certain style aspects of the payment buttons may be controlled via a remote setting outside of the regular plugin settings. Otherwise, it may cause confusion if settings configurations are not adapted to this and e.g. the settings are configured with a button height of `25px`, but then on the block pages, it gets overridden to `55/56px` causing a mismatch in expectations. 
PayPal Payments currently has a `Customize smart button style per location` setting which we intended to rework into a `Use PayPal's recommended button styling` setting that would _disable_ certain configuration options from the settings in favor of a pre-defined default configuration. Maybe a setting like this that actively removes certain configuration options from the plugin level while enabled could be combined with the efforts of this proposal to clearly communicate with users from where their button appearance is controlled.

**Reply by alexflorisca** [reply]
> Because the Large style is defined with 56px, but PayPal only supports up to 55px for the buttons

Turns out the new [Stripe Express Checkout Elements](https://docs.stripe.com/elements/express-checkout-element) API also supports up to a height of max 55px. So we can change the max height here to align with both PayPal and Stripe's requirements.

> Beyond this, what I am missing from the proposal is a general guideline on how the payment gateway configuration should communicate with users that certain style aspects of the payment buttons may be controlled via a remote setting outside of the regular plugin settings.

This is something that was [mentioned above](https://github.com/woocommerce/woocommerce/discussions/48055#discussioncomment-9648533) and is something I will focus on next. My immediate thoughts here are that we should provide good defaults for both blocks and non blocks context, and make it clear to the user which settings apply where. This applies Both in the express checkout block where they can modify "common settings" and in the payment gateway's configuration page where they can modify gateway specific settings. I'll give this some more thought over the coming days

**Reply by pmcpinto** [reply]
Niklas, thanks for your feedback.

> Beyond this, what I am missing from the proposal is a general guideline on how the payment gateway configuration should communicate with users that certain style aspects of the payment buttons may be controlled via a remote setting outside of the regular plugin settings. Otherwise, it may cause confusion if settings configurations are not adapted to this and e.g. the settings are configured with a button height of 25px, but then on the block pages, it gets overridden to 55/56px causing a mismatch in expectations.
>PayPal Payments currently has a Customize smart button style per location setting which we intended to rework into a Use PayPal's recommended button styling setting that would disable certain configuration options from the settings in favor of a pre-defined default configuration. Maybe a setting like this that actively removes certain configuration options from the plugin level while enabled could be combined with the efforts of this proposal to clearly communicate with users from where their button appearance is controlled.

Good point. cc @elizaan36 @nikkivias as this should be considered in the guidelines.

**Comment by elizaan36** [comment]
Great work @alexflorisca, exciting to see this progress and I agree with the properties you've suggested to set defaults for. @nikkivias Do you think it would help to share your proposal for the shared express button appearance settings on this thread? It's related to the technical discussion of creating strong defaults, and it may spark a conversation on what aspects of the visual appearance merchants could have control over. Smart defaults are needed, but the payment buttons also need to align on the variations offered.

**Comment by ralucaStan** [comment]
> @ralucaStan made a good suggestion about removing the label altogether and just displaying the default button logo. 

In the meantime, I've reconsidered my suggestion because that label is needed for a11y purposes, namely for Screen readers. 

When provided, it should still be present in the markup, even if not visually present. We'd need to see what is the best way to represent this information to screen readers: hidden text + icon in the button, an aria-label tag etc

**Reply by alexflorisca** [reply]
> When provided, it should still be present in the markup, even if not visually present. We'd need to see what is the best way to represent this information to screen readers: hidden text + icon in the button, an aria-label tag etc

This is provided by each payment extension that renders the button. I've checked PayPal, WooPay and Stripe and they all have great support for voice over. Without any label, the voice over reads "Google Pay, buttoin", "WooPay, button", "PayPal, button". So I think that is good enough as a default experience.

Like I said, my thinking now is that we should default for no label on all exprsss payment methods (just the logo or payment method name), with the option to add a label for each one individually

**Reply by pierorocca** [reply]
> with the option to add a label for each one individually
Question to @nikkivias  and @elizaan36  is there any value in enabling label customization on a per page basis, defaulting to no label in all cases?

**Reply by lauramelos** [reply]
hey @alexflorisca here are our comments regarding Amazon Pay:

> Does this approach work with your payment extension?

We can use the new height attribute and use it in the button’s container. As you stated in the discussion (n/a) we cannot control the corner radius or the button label).
Could you add some logic to change the button color given the dark mode setting?
Amazon pay button supports the following color schemes: Gold, Light Grey, and Dark Grey.
Passing the proper color to the button is an easy task, but we should leave the merchant to decide which scheme will be rendered on light and dark themes.
An initial approach would be to add a couple of settings to the gateway settings or the theme’s customizer so the merchant can decide which color scheme will be displayed in light and dark mode. The main reason behind this is that the gold button looks good either in light or dark mode.

> Do you see any problems or conflicts with this approach?

We don’t envision any problem or conflict with this approach

> Is there a better solution?

For the case of Amazon Pay we render the button directly from amazon api response, but we could check with Amazon if it posible to change the current implementation (at least for the cart/checkout blocks to minimize any risk of breaking the UX for merchants that are using the current integration). We could render a normal HTML <button>, so we can control the aspect of the button. 
We would init the amazon checkout whenever the customer clicks the button, that way we can ensure that we control all the aspects of the button’s appearance (label, icon, color, etc) form the plugin itself and still be compatible with the Amazon Pay flow.
Again: this is conditional to Amazon answer, in current escenario is not possible. Will update when we have it.

**Reply by alexflorisca** [reply]
Thanks @lauramelos for your reply. 

> We can use the new height attribute and use it in the button’s container.

That's great. The height is the most important attribute for the button to respect. But you make a good point here and we could explore setting a max-height property on the grid elements where the buttons are rendered which would contain them. I guess the buttons would need to either render at the same height, or height: 100% which would take up the entire container. I'm going to explore this a little further actually, maybe we can avoid passing the height property alltogether and have all buttons render `height: 100%` instead. 

> Passing the proper color to the button is an easy task, but we should leave the merchant to decide which scheme will be rendered on light and dark themes.

Agreed. If the default Gold color looks good on the majority of themes (light and dark) then we don't need any custom logic to change the color. The idea was to use this `darkMode` variable for buttons that render colors  which only work in either light or dark themes.

> An initial approach would be to add a couple of settings to the gateway settings or the theme’s customizer so the merchant can decide which color scheme will be displayed in light and dark mode. The main reason behind this is that the gold button looks good either in light or dark mode.

I think the only thing that's necessary is a setting for the user to change the color for the current theme. I don't think we need to go to the detail of setting specific colors for light vs dark themes. That may already exist, I can't remember.

> We don’t envision any problem or conflict with this approach

Great news!

> We could render a normal HTML , so we can control the aspect of the button.

That would be great if it's possible. The label is not such a big deal, I think we're going to default to "no label" (i.e. just the name of the payment gateway). And if the merchant wants to add any extra label to it (e.g. "Buy with"), they will do this from the payment gateway settings for each payment button. So it would be a bonus if this was possible, but not a necessity. 

The border radius would be more important here in order to align the button styles. This will be a more noticeable difference to shoppers if one button has/hasn't got rounded corners and the others do. 

> Will update when we have it.
Appreciate that thank you. I'd love to wrap up this discussion by 17th June if possible. Look forward to your update here.

Thanks again!

**Reply by alexflorisca** [reply]
> I guess the buttons would need to either render at the same height, or height: 100% which would take up the entire container. I'm going to explore this a little further actually, maybe we can avoid passing the height property alltogether and have all buttons render height: 100% instead.

This isn't possible due to the restriction on PayPal and Stripe SDKs, requiring button height to be specified as an exact pixel value :(

**Comment by alexflorisca** [comment]
I wanted to address a missing part of this proposal:

> a general guideline on how the payment gateway configuration should communicate with users that certain style aspects of the payment buttons may be controlled via a remote setting outside of the regular plugin settings

With this proposal, we're introducing another way for merchants to customise the express payment buttons. This may cause a bit of confusion around which settings are applied when and in what context.

What I propose here is as follows:
- A toggle within the express checkout block that will allow merchants to override gateway specific settings with these new common settings.
- If this toggle is on, common settings such as _button height_ and _border radius_ will be displayed with the merchants able to select from these, much like in the POC linked above.
- If this toggle is off, a notice will be displayed informing the merchant that these settings are controlled by each payment gateway, with a link to the payment settings page in WooCommerce.
- An info message within each extension's settings area, that reads something like "Some of the appearance settings here may be overridden by the Express Checkout Block's settings" with a link to the checkout page where they can be configured

It's not the sexiest solution, but this proposed API will be very light (only height, darkMode and possibly border radius will be available as common settings across all buttons to start with). Given sensible defaults on both sides, I think this provides enough clarity to merchants. I would love to do something a bit more clever here, but I can't think of anything that works.

The alternative approach I considered is to remove some of the extension specific settings that are provided by the express checkout block. For example, the height would be removed from the gateway specific settings, and only live under the express checkout block. This doesn't really work because we're not always in a block context and still need to support classic cart & checkout, product pages and admin pages

Another approach would be to separate shortcode from block settings in each gateway's settings page. This could work but I think it would be even more confusing for the merchants as some settings would be within the gateway settings page (e.g. button label, which we can't generalise across extensions) and some would be in the express checkout block (height, border-radius). So we're still sending users to 2 different places. It's also tricky to enforce what each gateway should do with their settings page. So being too prescriptive here and relying on that is not a good idea in my opinion.

I think it's a little clearer to say, "here's the extension specific options, which can be overridden from the express checkout block." It's not perfect but I think it's better than the alternatives. @InpsydeNiklas what do you think given your ideas around re-working the custom styles per location setting in the PayPal extension?

**Reply by opr** [reply]
> A toggle within the express checkout block that will allow merchants to override gateway specific settings with these new common settings.

Would this be toggled on by default? I think the default experience should be one that makes everything the same.

**Reply by InpsydeNiklas** [reply]
As mentioned above, it makes sense to ship the Checkout block pre-configured with the override enabled and make it up to the payment integration to detect and apply custom rules for communicating this "external configuration" with the user. 

While removing/disabling certain settings would be possible (it is already actively used in the PayPal Payments plugin), it would introduce complexity early on, may impact documentation in various places, and has the potential to cause confusion among some users. Let's stick with keeping it simple.
> An info message within each extension's settings area, that reads something like "Some of the appearance settings here may be overridden by the Express Checkout Block's settings" with a link to the checkout page where they can be configured

The configurations across different payment plugins are too diverse for one single solution, but a strategically placed message with a somewhat consistent labeling can be produced in any integration. PayPal Payments recently introduced an "inline notice" to better communicate certain limitations. It mimics the visuals of regular admin notices while utilizing custom placements to not "drown in the noise of 10+ admin notices." 
This may be one possible way to communicate with users that certain settings are controlled from another place:
![image](https://github.com/woocommerce/woocommerce/assets/70433191/ef6b7d20-96f6-4ec9-af23-9934d8521e88)



For a start, I prefer to keep the settings as usual without hiding or disabling any specific setting and instead inform users about it with a high-visibility notice/banner.
Once the integration matures and potentially covers the majority of the button configurations (including language, label, etc.), it's worth re-evaluating how payment plugins may communicate this with users.

**Reply by alexflorisca** [reply]
> Would this be toggled on by default? I think the default experience should be one that makes everything the same.

Yep, this is what I intended as well, but forgot to mention

> PayPal Payments recently introduced an "inline notice" to better communicate certain limitations. It mimics the visuals of regular admin notices while utilizing custom placements to not "drown in the noise of 10+ admin notices."

I like that, we could start with that and like you say, tweak it further in the future if needs be. I also think the message is pretty spot on so we can use that as a starter. Would you be happy to add this to the [PR](https://github.com/woocommerce/woocommerce-paypal-payments/pull/2354) you did to integrate with this work? I think you're better placed to decide where and how this notice is best shown within the PayPal Payments app.

Also would this live alongside the work you're doing with the recommended button styling?

**Comment by pierorocca** [comment]
> Purple
Light Purple
Light Purple Outline

Note WooPay has moved away from the light purple to neutral white. I'm also interested in providing a Black alternative to the default Purple. It's not something that could be programmatically chosen. A merchant would need to choose it over the default option. In this proposal, where would that option reside?

**Comment by pierorocca** [comment]
FYI I've reached out to Stripe about the ECE 55px upper limit. The PRB element didn't have the same constraint so trying to understand why the difference and if it's an arbitrary number, could that increase that to 56px to avoid having 3Ps from having to update their sizing.

**Comment by pierorocca** [comment]
Apologies, I didn't include the response here. Stripe indicated that the 55px limitation is a result of PayPal and their max supported button size. PayPal is an option in the ECE while it is not in the PRB. So it's capped due to that 3P dependency.

**Comment by Sidsector9** [comment]
Hi @alexflorisca great work! I have tested this out with Square and works great. Since the express payment buttons can be displayed on the Block Checkout, Cart, and Single Product pages, does that mean the merchant will have to set the styling on all three blocks individually?

**Reply by alexflorisca** [reply]
Whenever the express payment buttons are rendered within a block context (so Cart and Checkout blocks), they should get the values from the new `buttonAttributes` API. Outside of a block context, they should respect the plugin specific settings. Can you share a link to where you tried this? 

I've recently created a PR in Square with a small fix and to try this out in [Add support for buttonAttributes and set default height to 48px](https://github.com/woocommerce/woocommerce-square/pull/175#top)
